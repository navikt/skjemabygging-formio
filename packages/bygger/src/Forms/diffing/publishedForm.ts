import featureToggles from "../../featureToggles";

async function loadPublishedForm(formPath) {
  if (featureToggles.enableDiff) {
    const response = await fetch(`/api/published-forms/${formPath}`, {
      method: "GET",
      headers: {
        "content-type": "application/json",
      },
    });
    if (response.ok) {
      return await response.json();
    }
  }
  return null;
}

export { loadPublishedForm };
