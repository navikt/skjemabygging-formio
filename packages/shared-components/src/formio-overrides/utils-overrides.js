import { formDiffingTool, navFormioUtils } from "@navikt/skjemadigitalisering-shared-domain";
import { Formio, Utils } from "formiojs";

Formio.Utils.toggleClass = (id, className) => {
  return `document.getElementById('${id}').classList.toggle('${className}')`;
};

Formio.Utils.getDiffLabel = (ctx) => {
  const { component, config } = ctx;
  const { publishedForm } = config;
  if (ctx.builder && publishedForm) {
    const changes = formDiffingTool.checkComponentDiff(component, publishedForm);
    if (changes && changes.status) {
      return `<span class="navds-tag navds-tag--warning-filled navds-tag--xsmall navds-detail navds-detail--small">${changes.status}</span>`;
    }
  }
  return "";
};

/*
 * This function is overridden because FormioUtils.sanitize calls dompurify.sanitize, which has a bug.
 * The bug is discussed here: https://github.com/cure53/DOMPurify/issues/276
 * In short, it reverses the order of attributes on HTML elements.
 *
 * For us, this breaks translations, as the text to be translated no longer matches the text in the translation file.
 *
 * By calling the sanitize-function twice, we reverse the reversed order, causing the original, correct order.
 * This is messy, but we consider it safer than turning off sanitizing by sending in "sanitize = false" to NavForm.
 */
const originalSanitize = Utils.sanitize;
Formio.Utils.sanitize = (string, options) => {
  return originalSanitize(originalSanitize(string, options), options);
};

const originalEvaluate = Utils.evaluate;

function evaluateOverride(func, args, ret, tokenize) {
  return originalEvaluate(sanitizeJavaScriptCode(func), args, ret, tokenize);
}

const { sanitizeJavaScriptCode } = navFormioUtils;

export { evaluateOverride, sanitizeJavaScriptCode };
