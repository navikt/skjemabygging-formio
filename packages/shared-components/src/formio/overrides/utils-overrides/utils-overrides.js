import fnrvalidator from '@navikt/fnrvalidator';
import { formDiffingTool, navFormioUtils } from '@navikt/skjemadigitalisering-shared-domain';
import { Utils } from 'formiojs';
import moment from 'moment/moment';
import '../moment-overrides';

const additionalDescription = (ctx) => {
  if (!ctx.component.additionalDescriptionLabel && !ctx.component.additionalDescriptionText) return '';

  const descriptionId = `${ctx.component.id}-${ctx.component.key}-additional-description`;
  const descriptionButtonId = `${ctx.component.id}-${ctx.component.key}-additional-button-content`;
  const descriptionContentId = `${ctx.component.id}-${ctx.component.key}-additional-description-content`;

  return `<div class="navds-read-more navds-read-more--medium" id="${descriptionId}">
    <button type="button" class="navds-read-more__button navds-body-short" aria-expanded="true" id="${descriptionButtonId}" onclick="(() => {          
      document.getElementById('${descriptionId}').classList.toggle('navds-read-more--open');          
      document.getElementById('${descriptionContentId}').classList.toggle('navds-read-more__content--closed');
      document.getElementById('${descriptionContentId}').setAttribute(
        'aria-hidden', (!document.getElementById('${descriptionContentId}').getAttribute('aria-hidden')).toString()
      );          
      document.getElementById('${descriptionButtonId}').setAttribute(
        'aria-expanded', (!document.getElementById('${descriptionButtonId}').getAttribute('aria-expanded')).toString()
      );
      })()">      
      <svg width="1em" height="1em" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg" focusable="false" role="img" class="navds-read-more__expand-icon" aria-hidden="true">
        <path fill-rule="evenodd" clip-rule="evenodd" d="M5.97 9.47a.75.75 0 0 1 1.06 0L12 14.44l4.97-4.97a.75.75 0 1 1 1.06 1.06l-5.5 5.5a.75.75 0 0 1-1.06 0l-5.5-5.5a.75.75 0 0 1 0-1.06Z" fill="currentColor"></path>
      </svg>
      <span>${ctx.t(ctx.component.additionalDescriptionLabel)}</span>
    </button>
    <div class="navds-read-more__content navds-read-more__content--closed" id="${descriptionContentId}" aria-hidden="true">
      ${ctx.t(ctx.component.additionalDescriptionText)}  
    </div>
  </div>`;
};

const translateHTMLTemplate = (template, translate) => {
  return translate(template);
};

const navFormDiffToHtml = (diffSummary) => {
  try {
    const { changesToCurrentComponent, deletedComponents } = diffSummary;
    const html = [];
    if (changesToCurrentComponent.length) {
      const labelId = 'nav-form-diff-changed-elements';
      html.push(`<span id="${labelId}" class="navds-body-short font-bold">Endringer</span>`);
      html.push(`<ul aria-labelledby="${labelId}">`);
      html.push(
        ...changesToCurrentComponent.map(
          (change) => `<li>${change.key}: Fra '${change.oldValue}' til '${change.newValue}'</li>`,
        ),
      );
      html.push('</ul>');
    }
    if (deletedComponents.length) {
      const labelId = 'nav-form-diff-deleted-elements';
      html.push(`<span id="${labelId}" class="navds-body-short font-bold">Slettede elementer</span>`);
      html.push(createList(deletedComponents, labelId));
    }
    return html.join('');
  } catch (err) {
    console.error(`Failed to render form diff: ${err.message} diffSummery="${JSON.stringify(diffSummary)}"`, err);
    return '<span>Det oppstod dessverre en feil under behandling av endringene i dette skjemaet.</span>';
  }
};

const createList = (components, labelId) => {
  if (components && components.length > 0) {
    const labelledBy = labelId ? ` aria-labelledby="${labelId}"` : '';
    return `<ul${labelledBy}>`
      .concat(
        components
          .map((component) => {
            return `<li>${component.type}: ${component.label}${createList(component.components)}</li>`;
          })
          .join(''),
      )
      .concat('</ul>');
  }
  return '';
};

const TAG = (text) =>
  `<span class="navds-tag navds-tag--warning navds-tag--xsmall navds-detail navds-detail--small">${text}</span>`;

const getDiffTag = (ctx) => {
  const { component, config, self } = ctx;
  const { publishedForm } = config;
  if (ctx.builder && publishedForm) {
    // Formio.js invokes mergeSchema on component which is put on ctx object. Therefore we must do the same
    // prior to comparing with published version to avoid misleading diff tags due to changes in a component's schema.
    const mergeSchema = self.mergeSchema.bind(self);
    const diff = formDiffingTool.getComponentDiff(component, publishedForm, mergeSchema);
    const tags = [];
    if (diff.isNew) {
      tags.push(`${TAG('Ny')}`);
    }
    if (diff.changesToCurrentComponent?.length) {
      tags.push(`${TAG('Endring')}`);
    }
    if (diff.deletedComponents?.length) {
      tags.push(`${TAG('Slettede elementer')}`);
    }
    return tags.join(' ');
  }
  return '';
};

/*
 * This function is overridden because FormioUtils.sanitize calls dompurify.sanitize, which has a bug.
 * The bug is discussed here: https://github.com/cure53/DOMPurify/issues/276
 * In short, it reverses the order of attributes on HTML elements.
 *
 * For us, this breaks translations, as the text to be translated no longer matches the text in the translation file.
 *
 * By calling the sanitize-function twice, we reverse the reversed order, causing the original, correct order.
 * This is messy, but we consider it safer than turning off sanitizing by sending in "sanitize = false" to NavForm.
 */
const originalSanitize = Utils.sanitize;
const sanitize = (string, options) => {
  return originalSanitize(originalSanitize(string, options), options);
};

const originalEvaluate = Utils.evaluate;

const evaluate = (func, args, ret, tokenize) => {
  return originalEvaluate(sanitizeJavaScriptCode(func), args, ret, tokenize);
};

const { sanitizeJavaScriptCode } = navFormioUtils;

const isBornBeforeYear = (year, fnrKey, submission = {}) => {
  const value = Utils.getValue(submission, fnrKey);
  if (value && fnrvalidator.fnr(value.trim()).status === 'valid') {
    const birthDateStr = value.substring(0, 6);
    return moment(birthDateStr, 'DDMMYY').year() < year;
  }
  return false;
};

const UtilsOverrides = {
  additionalDescription,
  translateHTMLTemplate,
  evaluate,
  sanitize,
  sanitizeJavaScriptCode,
  navFormDiffToHtml,
  getDiffTag,
  isBornBeforeYear,
};

if (typeof global === 'object' && global.FormioUtils) {
  Object.entries(UtilsOverrides).forEach(([key, value]) => {
    Utils[key] = value;
    // FormioUtils is set on global scope by Formio if global is object
    global.FormioUtils[key] = value;
  });
}

// Formio require lodash to be available on window for custom conditionals.
if (!!Utils._ && !window._) {
  window._ = Utils._;
}

export default UtilsOverrides;
