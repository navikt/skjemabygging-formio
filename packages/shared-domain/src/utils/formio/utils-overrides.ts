import _ from 'lodash';
import sanitizeJavaScriptCode from './sanitize-javascript-code';

const getOverrides = (Utils) => {
  /*
   * This function is overridden because FormioUtils.sanitize calls dompurify.sanitize, which has a bug.
   * The bug is discussed here: https://github.com/cure53/DOMPurify/issues/276
   * In short, it reverses the order of attributes on HTML elements.
   *
   * For us, this breaks translations, as the text to be translated no longer matches the text in the translation file.
   *
   * By calling the sanitize-function twice, we reverse the reversed order, causing the original, correct order.
   * This is messy, but we consider it safer than turning off sanitizing by sending in "sanitize = false" to NavForm.
   */
  const originalSanitize = Utils.sanitize;
  const sanitize = (string, options) => {
    return originalSanitize(originalSanitize(string, options), options);
  };

  const originalEvaluate = Utils.evaluate;
  const evaluate = (func, args, ret?, tokenize?) => {
    return originalEvaluate(sanitizeJavaScriptCode(func), args, ret, tokenize);
  };

  // override to skip formio's default HTML translation
  const translateHTMLTemplate = (template, translate) => {
    return translate(template);
  };

  /* Override to ensure that utils and submission are passed to evaluate.
   * These are automagically passed by Formio when the form is rendered using templates,
   * but not when checkCondition is invoked manually, e.g., when formSummaryUtils is used in
   * fyllut-backend, or when rendering FormStepper (React).
   */
  const checkCustomConditional = (component, custom, row, data, form, variable, onError, instance, submission) => {
    if (typeof custom === 'string') {
      custom = `var ${variable} = true; ${custom}; return ${variable};`;
    }
    const evaluateArgs = { row, data, form, utils: Utils, ...(submission && { submission }), _ };
    const value =
      instance && instance.evaluate ? instance.evaluate(custom, evaluateArgs) : evaluate(custom, evaluateArgs);
    if (value === null) {
      return onError;
    }
    return value;
  };

  /*
   * Override to invoke checkCustomConditional with submission, which is needed for miscellaneous util functions
   */
  const originalCheckCondition = Utils.checkCondition;
  const checkCondition = (component, row, data, form, instance, submission) => {
    const { customConditional } = component;
    if (customConditional) {
      return checkCustomConditional(component, customConditional, row, data, form, 'show', true, instance, submission);
    }
    return originalCheckCondition(component, row, data, form, instance);
  };

  return {
    sanitize,
    evaluate,
    translateHTMLTemplate,
    checkCustomConditional,
    checkCondition,
  };
};

export default getOverrides;
