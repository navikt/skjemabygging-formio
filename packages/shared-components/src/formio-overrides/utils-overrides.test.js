import { formDiffingTool, navFormUtils } from "@navikt/skjemadigitalisering-shared-domain";
import panelDiffDeletedDatagrid from "./testdata/diff-deleted-datagrid";
import panelDiffDeletedRadiopanel from "./testdata/diff-deleted-radio";
import formNavSelectChanges from "./testdata/form-navSelect-changes";
import publishedForm from "./testdata/published-form";
import { getDiffTag, navFormDiffToHtml, sanitizeJavaScriptCode } from "./utils-overrides";

describe("utils-overrides", () => {
  describe("sanitizeJavaScriptCode", () => {
    it("does not change a string without chained lookups", () => {
      const aSentence = "this is a string.";
      expect(sanitizeJavaScriptCode(aSentence)).toEqual(aSentence);

      const aRealisticInputWithoutChainedLookups = "show = a === 'b'";
      expect(sanitizeJavaScriptCode(aRealisticInputWithoutChainedLookups)).toEqual(
        aRealisticInputWithoutChainedLookups
      );
    });

    it("correctly adds null/undefined checks for chained lookups", () => {
      const inputWithChainedLookups = "show = a.b === 'c'";
      expect(sanitizeJavaScriptCode(inputWithChainedLookups)).toEqual("show = (a && a.b) === 'c'");
    });

    it("correctly adds null/undefined checks for multiple chained lookups", () => {
      const inputWithMultipleChainedLookups = "show = a.b === 'c' || d.e === 'f'";
      expect(sanitizeJavaScriptCode(inputWithMultipleChainedLookups)).toEqual(
        "show = (a && a.b) === 'c' || (d && d.e) === 'f'"
      );
    });

    it("correctly adds null/undefined checks for deeply chained lookups", () => {
      const inputWithDeeplyChainedLookups = "show = a.b.c.d.e === 'f'";
      expect(sanitizeJavaScriptCode(inputWithDeeplyChainedLookups)).toEqual(
        "show = (a && a.b && a.b.c && a.b.c.d && a.b.c.d.e) === 'f'"
      );
    });

    it("correctly add null/undefined checks for multiple equal chained lookups", () => {
      const inputWithMultipleEqualChainedLookups = "show = a.b === 'c' || a.b === 'd'";
      expect(sanitizeJavaScriptCode(inputWithMultipleEqualChainedLookups)).toEqual(
        "show = (a && a.b) === 'c' || (a && a.b) === 'd'"
      );
    });

    it("correctly add null/undefined checks when variable names includes numbers", () => {
      const inputWithMultipleEqualChainedLookups = "show = a1.b2 === 'c'";
      expect(sanitizeJavaScriptCode(inputWithMultipleEqualChainedLookups)).toEqual("show = (a1 && a1.b2) === 'c'");
    });

    it("will not change a partial expression", () => {
      const inputWithTwoChainedWhereOneIsAPartialOfTheOther =
        "show = anObject.aString === 'c' || anObject.aString1 === 'd'";
      const actual = sanitizeJavaScriptCode(inputWithTwoChainedWhereOneIsAPartialOfTheOther);
      expect(actual).toEqual(
        "show = (anObject && anObject.aString) === 'c' || (anObject && anObject.aString1) === 'd'"
      );
    });

    it("will not change a partial expression that ends in an equal expression to another complete expression", () => {
      const inputWithOneChainedExpressionEndingInAnotherExpression =
        "show = anObject.aString === 'c' || Object.aString === 'd'";
      const actual = sanitizeJavaScriptCode(inputWithOneChainedExpressionEndingInAnotherExpression);
      expect(actual).toEqual("show = (anObject && anObject.aString) === 'c' || (Object && Object.aString) === 'd'");
    });

    describe("When the code includes function calls", () => {
      it("does not add null checks for functions on instance", () => {
        const inputWithInstanceFunctionCall = "valid = instance.validate(input)";
        expect(sanitizeJavaScriptCode(inputWithInstanceFunctionCall)).toEqual("valid = instance.validate(input)");
      });

      it("does not add null checks for functions on util", () => {
        const inputWithUtilFunctionCall = "valid = util.fun(input)";
        expect(sanitizeJavaScriptCode(inputWithUtilFunctionCall)).toEqual("valid = util.fun(input)");
      });

      it("does not add null checks for functions on utils", () => {
        const inputWithUtilsFunctionCall = "valid = utils.fun(input)";
        expect(sanitizeJavaScriptCode(inputWithUtilsFunctionCall)).toEqual("valid = utils.fun(input)");
      });

      it("does not add null checks for functions on lodash", () => {
        const inputWithLodashFunctionCall = "valid = _.fun(input)";
        expect(sanitizeJavaScriptCode(inputWithLodashFunctionCall)).toEqual("valid = _.fun(input)");
      });

      it("does not add null checks for nested functions on a reserved word", () => {
        const inputWithNestedFunctionCalls = "valid = instance.fun1(_.some(data, (a) => util.fun2(a.b.c)))";
        expect(sanitizeJavaScriptCode(inputWithNestedFunctionCalls)).toEqual(
          "valid = instance.fun1(_.some(data, (a) => util.fun2((a && a.b && a.b.c))))"
        );
      });

      it("does not add null checks for function calls", () => {
        const inputWithFunctionCall = "valid = obj.myFunction()";
        expect(sanitizeJavaScriptCode(inputWithFunctionCall)).toEqual("valid = obj.myFunction()");

        const inputWithFunctionThatTakesParams = "valid = obj.myFunction(someInput)";
        expect(sanitizeJavaScriptCode(inputWithFunctionThatTakesParams)).toEqual("valid = obj.myFunction(someInput)");

        const inputWithNestedFunctionCall = "valid = parentObject.childObject.nestedFunction()";
        expect(sanitizeJavaScriptCode(inputWithNestedFunctionCall)).toEqual(
          "valid = parentObject.childObject.nestedFunction()"
        );

        const inputWithNestedObjectReferenceAndFunctionCall = "valid = obj.someVar || obj.someFunction()";
        expect(sanitizeJavaScriptCode(inputWithNestedObjectReferenceAndFunctionCall)).toEqual(
          "valid = (obj && obj.someVar) || obj.someFunction()"
        );
      });
    });
  });

  describe("navFormDiffToHtml", () => {
    it("generates html list with changed properties", () => {
      const changes = {
        status: "Endring",
        diff: {
          label: { status: "Endring", originalValue: "Fornavn", value: "Oppgi fornavn" },
          id: { status: "Endring", originalValue: "ehnemzu", value: "e6s77h" },
        },
        key: "fornavn",
        type: "textfield",
      };
      const diffSummary = formDiffingTool.createDiffSummary(changes);
      const html = navFormDiffToHtml(diffSummary);
      expect(html).toMatchSnapshot();
    });

    it("generates html list with deleted radiopanel from panel", () => {
      const diffSummary = formDiffingTool.createDiffSummary(panelDiffDeletedRadiopanel);
      const html = navFormDiffToHtml(diffSummary);
      expect(html).toMatchSnapshot();
    });

    it("generates nested html list with deleted datagrid and its components", () => {
      const diffSummary = formDiffingTool.createDiffSummary(panelDiffDeletedDatagrid);
      const html = navFormDiffToHtml(diffSummary);
      expect(html).toMatchSnapshot();
    });

    it("handles failure when diffSummary is of unexpected type", () => {
      const originalConsoleError = console.error;
      console.error = () => {};
      try {
        const diffSummary = { some: "thing", is: "wrong", with: ["this", "obj", "ect"] };
        const html = navFormDiffToHtml(diffSummary);
        expect(html).toMatchSnapshot();
      } catch (e) {
        throw new Error(`Should never fail, but it did: ${e.message}`);
      } finally {
        console.error = originalConsoleError;
      }
    });

    describe("Form -> diff -> html", () => {
      it("empty html when navSelect is not changed", () => {
        const navSelect = navFormUtils.findByNavId("e0a8kbj", publishedForm.components);
        const componentDiff = formDiffingTool.getComponentDiff(navSelect, publishedForm);
        const html = navFormDiffToHtml(componentDiff);
        expect(html).toEqual("");
        expect(html).toMatchSnapshot();
      });
      it("should list changes for navSelect", () => {
        const navSelect = navFormUtils.findByNavId("e0a8kbj", formNavSelectChanges.components);
        const componentDiff = formDiffingTool.getComponentDiff(navSelect, publishedForm);
        const html = navFormDiffToHtml(componentDiff);
        expect(html).toMatchSnapshot();
      });
      it("should list changed key", () => {
        const radiopanel = navFormUtils.findByKey("doYouLiveInNorway", formNavSelectChanges.components);
        const componentDiff = formDiffingTool.getComponentDiff(radiopanel, publishedForm);
        const html = navFormDiffToHtml(componentDiff);
        expect(html).toMatchSnapshot();
      });
      it("should list change in conditional", () => {
        const alertstripe = navFormUtils.findByKey("alertstripeArstid", formNavSelectChanges.components);
        const componentDiff = formDiffingTool.getComponentDiff(alertstripe, publishedForm);
        const html = navFormDiffToHtml(componentDiff);
        expect(html).toMatchSnapshot();
      });
    });
  });

  describe("getDiffTag", () => {
    it("returns no tag when component is unchanged", () => {
      const navSelect = navFormUtils.findByNavId("e0a8kbj", publishedForm.components);
      const ctx = {
        builder: true,
        component: navSelect,
        config: {
          publishedForm,
        },
      };
      const html = getDiffTag(ctx);
      expect(html).toMatchSnapshot();
    });

    it("returns tag 'Endring' when component is changed", () => {
      const navSelect = navFormUtils.findByNavId("e0a8kbj", formNavSelectChanges.components);
      const ctx = {
        builder: true,
        component: navSelect,
        config: {
          publishedForm,
        },
      };
      const html = getDiffTag(ctx);
      expect(html).toMatchSnapshot();
    });

    it("returns no tag even if component is changed when it renders outside builder", () => {
      const navSelect = navFormUtils.findByNavId("e0a8kbj", formNavSelectChanges.components);
      const ctx = {
        builder: false,
        component: navSelect,
        config: {
          publishedForm,
        },
      };
      const html = getDiffTag(ctx);
      expect(html).toMatchSnapshot();
    });

    it("returns tag 'Ny' when component does not exist in published form", () => {
      const component = { navId: "123456", key: "aaaa" };
      const ctx = {
        builder: true,
        component,
        config: {
          publishedForm,
        },
      };
      const html = getDiffTag(ctx);
      expect(html).toMatchSnapshot();
    });
  });
});
