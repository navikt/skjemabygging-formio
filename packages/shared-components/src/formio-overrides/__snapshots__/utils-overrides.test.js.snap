// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils-overrides navFormDiffToHtml Form -> diff -> html empty html when navSelect is not changed 1`] = `""`;

exports[`utils-overrides navFormDiffToHtml Form -> diff -> html should list change in conditional 1`] = `"<span id=\\"nav-form-diff-changed-elements\\" class=\\"skjemaelement__label\\">Endringer</span><ul aria-labelledby=\\"nav-form-diff-changed-elements\\"><li>conditional.eq: Fra 'sommer' til 'vinter'</li></ul>"`;

exports[`utils-overrides navFormDiffToHtml Form -> diff -> html should list changed key 1`] = `"<span id=\\"nav-form-diff-changed-elements\\" class=\\"skjemaelement__label\\">Endringer</span><ul aria-labelledby=\\"nav-form-diff-changed-elements\\"><li>key: Fra 'borDuINorge1' til 'doYouLiveInNorway'</li></ul>"`;

exports[`utils-overrides navFormDiffToHtml Form -> diff -> html should list changes for navSelect 1`] = `"<span id=\\"nav-form-diff-changed-elements\\" class=\\"skjemaelement__label\\">Endringer</span><ul aria-labelledby=\\"nav-form-diff-changed-elements\\"><li>label: Fra 'Instrument' til 'Velg instrument'</li><li>data.values: Fra '[{\\"label\\":\\"Gitar\\",\\"value\\":\\"gitar\\"},{\\"label\\":\\"Trommer\\",\\"value\\":\\"trommer\\"},{\\"label\\":\\"Piano\\",\\"value\\":\\"piano\\"},{\\"label\\":\\"Trekkspill\\",\\"value\\":\\"trekkspill\\"},{\\"label\\":\\"Munnspill\\",\\"value\\":\\"munnspill\\"}]' til '[{\\"label\\":\\"Gitar\\",\\"value\\":\\"gitar\\"},{\\"label\\":\\"Trommer\\",\\"value\\":\\"trommer\\"},{\\"label\\":\\"Piano\\",\\"value\\":\\"piano\\"},{\\"label\\":\\"Trekkspill\\",\\"value\\":\\"trekkspill\\"}]'</li><li>defaultValue: Fra 'undefined' til '{\\"label\\":\\"Trommer\\",\\"value\\":\\"trommer\\"}'</li></ul>"`;

exports[`utils-overrides navFormDiffToHtml generates html list with changed properties 1`] = `"<span id=\\"nav-form-diff-changed-elements\\" class=\\"skjemaelement__label\\">Endringer</span><ul aria-labelledby=\\"nav-form-diff-changed-elements\\"><li>label: Fra 'Fornavn' til 'Oppgi fornavn'</li><li>id: Fra 'ehnemzu' til 'e6s77h'</li></ul>"`;

exports[`utils-overrides navFormDiffToHtml generates html list with deleted radiopanel from panel 1`] = `"<span id=\\"nav-form-diff-deleted-elements\\" class=\\"skjemaelement__label\\">Slettede elementer</span><ul aria-labelledby=\\"nav-form-diff-deleted-elements\\"><li>radiopanel: Bekreftelse p√• skoleplass</li></ul>"`;

exports[`utils-overrides navFormDiffToHtml generates nested html list with deleted datagrid and its components 1`] = `"<span id=\\"nav-form-diff-deleted-elements\\" class=\\"skjemaelement__label\\">Slettede elementer</span><ul aria-labelledby=\\"nav-form-diff-deleted-elements\\"><li>datagrid: Dine hunder<ul><li>textfield: Navn</li><li>textfield: Rase</li></ul></li></ul>"`;

exports[`utils-overrides navFormDiffToHtml handles failure when diffSummary is of unexpected type 1`] = `"<span>Det oppstod dessverre en feil under behandling av endringene i dette skjemaet.</span>"`;
